phylo/acceleration.go

```diff
@@ -0,0 +1,339 @@
+	WchimpGor   float64
+	WchimpOrang float64
+	WgorOrang   float64
+}
+
+//AccelBranchLengths represents the set of branch lengths corresponding to a particular distance matrix used for acceleration analysis on a 4-species alingment.
+type AccelBranchLengths struct {
+	BhumHca   float64
+	BchimpHca float64
+	BhcaHga   float64
+	BhgaGor   float64
+	BhgaOrang float64
+}
+
+//AccelSubTreeLeft is a tree with three leaves and three branches, joined at the single internal node.
+//Dij represent the observed pairwise distances between two species. vi represents the length of the branch between a species i and the internal node at the current stage of optimization.
+//Wij are the weight constants corresponding to each distance Dij such that Wij = 1 / (Dij^2)
+type AccelSubTreeLeft struct {
+	DhumChimp float64
+	DhumHga   float64
+	DchimpHga float64
+	VhumHca   float64
+	VchimpHca float64
+	VhcaHga   float64
+	WhumChimp float64
+	WhumHga   float64
+	WchimpHga float64
+}
+
+//AccelSubTreeRight is a tree with three leaves and three branches, joined at the single internal node.
+//Dij represent the observed pairwise distances between two species. vi represents the length of the branch between a species i and the internal node at the current stage of optimization.
+//Wij are the weight constants corresponding to each distance Dij such that Wij = 1 / (Dij^2)
+type AccelSubTreeRight struct {
+	DgorOrang float64
+	DhcaGor   float64
+	DhcaOrang float64
+	VhcaHga   float64
+	VhgaGor   float64
+	VhgaOrang float64
+	WgorOrang float64
+	WhcaGor   float64
+	WhcaOrang float64
+}
+
+//BranchLengthsAlternatingLeastSquares calculates the optimal branch lengths for a given set of distances. See the multiFaAcceleration readme for a detailed theoretical description of this algorithm.
+func BranchLengthsAlternatingLeastSquares(d AccelDistancesAndWeights, allowNegative bool, verbose bool, zeroDistanceWeightConstant float64, epsilon float64) AccelBranchLengths {
+	var answer = AccelBranchLengths{1, 1, 1, 1, 1}
+	var Q float64 = calculateQ(d, answer)
+	var nextQ float64
+	var currDiff float64 = epsilon + 1 //set currDiff to something larger than epsilon so that we make it into the loop the first time.
+	var leftSub AccelSubTreeLeft
+	var rightSub AccelSubTreeRight
+	var maxIteration, i = 1000, 0
+	var oldAnswer = AccelBranchLengths{1, 1, 1, 1, 1}
+
+	for currDiff > epsilon && i < maxIteration {
+		oldAnswer = answer
+		pruneLeft(d, answer, &leftSub, zeroDistanceWeightConstant)
+		optimizeSubtreeLeft(&leftSub, allowNegative, verbose, &answer)
+		pruneRight(d, answer, &rightSub, zeroDistanceWeightConstant)
+		optimizeSubtreeRight(&rightSub, allowNegative, verbose, &answer)
+		nextQ = calculateQ(d, answer)
+		//DEBUG: log.Printf("nextQ: %e. currDiff: %e. Here were the branch lengths: %f. %f. %f. %f. %f.", nextQ, currDiff, answer.B1, answer.B2, answer.B3, answer.B4, answer.B5)
+		currDiff = math.Abs(Q - nextQ)
+		if nextQ > Q { //nextQ is higher than Q, which means we got "worse"
```

Comment:
Does this still happen with the bug fix?

Reply:
unclear, I'd have to run it genome wide with a verbose check here